### PR TITLE
fixed git pull

### PR DESCRIPTION
**Pull request by cosminbasca**
Changed the behavior of GitConfig.pull. 
Original behavior: when update=True, do a git pull which works well when the refspec is not a commit hash.
New behavior: when update=True and rev not None, delegate to GitConfig.udpate (with force=True, discard local changes). Should this behavior be different? 

Bonus, updated GitConfig.update to suppress the "detached HEAD" warning messages when not in verbose mode. 

Bumped version to 5.6.15

### CODE REVIEW

File: mx.py [hunk 1]
```diff
@@ -3445,13 +3445,17 @@ def pull(self, vcdir, rev=None, update=False, abortOnError=True):
...
-            out = OutputCapture()
-            rc = self.run(cmd, nonZeroIsFatal=abortOnError, cwd=vcdir, out=out)
-            logvv(out.data)
-            return rc == 0
+            if rev:
+                return self.update(vcdir, rev=rev, mayPull=False, clean=True, abortOnError=abortOnError)
+            else:
+                # refspec = '{0}:{1}'.format(rev if rev else 'HEAD', active_branch)
+                refspec = 'HEAD:{0}'.format(active_branch)
+                cmd = ['git', 'pull', 'origin', refspec]
```
Comment (dougxc) [inline mx.py]: @gilles-duboscq stated that it seems wrong to execute a `pull` after having executed `fetch` given that `pull` also does a `fetch`. I'd like to get his input of this change before merging it.

Comment (chumer) [inline mx.py]: I see but. I desperately need this change as graal-vm is broken due to this.
Doing a pull after a fetch is definitely not what we wnat if a revision is specified.

Comment (cosminbasca) [inline mx.py]: if I understand correctly, the fetch should only occur on the code branch where we do an update. If we go with the pull we don't do a fetch. Is this correct?

File: mx.py [hunk 2]
```diff
@@ -3445,13 +3445,17 @@ def pull(self, vcdir, rev=None, update=False, abortOnError=True):
...
-            cmd = ['git', 'pull', 'origin', refspec]
-            self._log_pull(vcdir, rev)
-            out = OutputCapture()
-            rc = self.run(cmd, nonZeroIsFatal=abortOnError, cwd=vcdir, out=out)
-            logvv(out.data)
-            return rc == 0
+            if rev:
+                return self.update(vcdir, rev=rev, mayPull=False, clean=True, abortOnError=abortOnError)
+            else:
+                # refspec = '{0}:{1}'.format(rev if rev else 'HEAD', active_branch)
```
Comment (dougxc) [inline mx.py]: Is there a good reason to leave this commented code in?

Comment (cosminbasca) [inline mx.py]: no, the comment can be removed

Comment (dougxc) [inline mx.py]: Ok, can you please remove it.

File: mx.py (hunk 1 above)
Comment (dougxc) [inline mx.py]: Sounds like the right thing to me.

**Comment by chumer** [comment]
I am merging because it was broken anyway. But @gilles-duboscq can you have another look? Thanks.

File: mx.py [hunk 3]
```diff
@@ -3442,20 +3442,23 @@ def pull(self, vcdir, rev=None, update=False, abortOnError=True):
...
             logvv(out.data)
             return rc == 0
         else:
-            if abortOnError:
-                abort('fetch returned ' + str(rc))
-            return False
+            rc = self._fetch(vcdir, abortOnError)
+            if rc == 0:
+                if rev:
+                    return self.update(vcdir, rev=rev, mayPull=False, clean=True, abortOnError=abortOnError)
```
Comment (gilles-duboscq) [inline mx.py]: I'm not sure about the `clean=True` that looks dangerous.
In the `hg` implementation we would not clean and rather abort if we can not update cleanly.

Also we should only update if `update` is true

**Pull request merged**